doc/man/man1/xdelta.1
\"    This documentation is copyright 1994 David Huron.
.TH xdelta 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +12
.ti -12
\fBxdelta\fR  --  calculate sequential numeric differences between successive Humdrum data tokens
.in -12
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
.in +7
.ti -7
\fBxdelta\fR  [\fB-ae\fR]  [\fB-b \fIregexp\fR]  [\fB-s \fIregexp\fR]  [\fIinputfile ...\fR]
.in -7
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "xdelta"
command calculates the numeric differences between successive
data tokens within individual spines.
By way of illustration,
.B "xdelta"
will change a sequence of numerical tokens -- such as
.I "x, y, z"
-- to their successive differences
.I "y\(mix, z\(miy"
(i.e.
.I "\(*Dx)."
The
.B "xdelta"
command might be used for such purposes as determining the melodic
interval distances between successive pitches in a musical line,
or calculating changes of duration for successive notes.
.sp 1
.sp 1
Each output interpretation is automatically assigned a new name by
prepending the letter `X' to the input interpretations.
For example, an input of
**semits
will result in an output interpretation named
**Xsemits.
Output from
.B "xdelta"
may be reprocessed as input, so that the `differences between the differences'
(i.e. second derivative) may be calculated.
Once again, the letter `X' will be prepended to the input interpretation,
so that an input interpretation such as
**Xfreq
will result in an output interpretation named
**XXfreq.
.sp 1
.sp 1
The
.B "xdelta"
command has a number of subtleties in its operation that facilitate
the processing of music-related data.
However, these subtleties can lead to unexpected results for the
inexperienced user;
caution is advised.
.sp 1
.sp 1
Numeric data are processed by calculating the arithmetic difference
between successive data tokens within each spine.
Hence, an input token `\f(CR\(mi5\fR' followed by `\f(CR\(mi2\fR'
will result in an output of `\f(CR3\fR'.
Positive output values indicate increasing input values;
negative output values indicate decreasing input values.
Null tokens are simply echoed in the output, however numerical
processing continues as if the null tokens were absent.
For example, the input token `\f(CR\(mi5\fR' followed by a null token,
followed by `\f(CR\(mi2\fR' will result in an output of `\f(CR3\fR'.
.sp 1
.sp 1
No difference value is calculated for the first numeric input token;
however, the numeric value of the first numeric token is echoed in the
output -- appearing in square brackets.
These initial values are referred to as
.I "offset"
values, since they indicate the starting value from which subsequent
differences are calculated.
Offset values can prove useful in attempting to reconstruct the input,
but the user may wish to eliminate offset values in subsequent processing
(see below).
.sp 1
.sp 1
All data tokens containing only non-numeric data are simply echoed in the
output.
In the default operation, data tokens without numbers cause the
difference calculation to be suspended, and the next occurrence of
a numeric token is treated as a new offset value for subsequent calculations.
As noted, null-tokens (.) are also directly echoed.
If data tokens contain both numeric and non-numeric data (e.g. \f(CR42abc\fR),
the default operation is to suppress the echoing of non-numeric signifiers in
the corresponding output.
Hence, an input token `\f(CR33b\fR' followed by `\f(CRp51.3xm\fR'
will result in the output token `\f(CR18.3\fR' (i.e. 51.3 \(mi 33).
However, the accompanying non-numeric data can be echoed in
the output by invoking the
.B "-e"
option.
.sp 1
.sp 1
The
.B "xdelta"
command is able to calculate unsigned (absolute) values where appropriate
-- using the
.B "-a"
option.
It is also able to handle multiple stops and data-flow interruptions
such as the occurrence of barlines.
By defining
.I "regular expression"
patterns, the user may select
which types of data tokens should be ignored by
.B "xdelta."
(See EXAMPLES below.)
.sp 1
.sp 1
Note that the output spine generated by
.B "xdelta"
preserves the same record-type structure as the input,
and so may readily be pasted with the input file using the Humdrum
.B "assemble"
command.
.in -2
.sp 1
.sp 1
.SH "OPTIONS"
.in +2
The
.B "xdelta"
command provides the following options:
.sp 1
.TS
l l.
\fB-a\fR	output absolute difference values
\fB-b \fIregexp\fR	break; do not calculate difference for tokens
	  matching \fIregexp;\fR restart difference
	  calculations with next numerical token
\fB-e\fR	echo non-numeric data for input tokens containing
	  numbers
\fB-h\fR	displays a help screen summarizing the command syntax
\fB-s \fIregexp\fR	skip; completely ignore tokens matching
	  \fIregexp;\fR (echo in output only)
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
The \(odskip\(cd function takes precedence over the \(odbreak\(cd function,
so input strings matching both the skip (-s) and break (-b) regular
expressions cause a skip rather than a break.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The various aspects of the
.B "xdelta"
command are best illustrated using a set of examples.
Consider the following input:
.sp 1
.sp 1
.in +2
.sp 1
.TS
l.
**cents
100
300
\.
1200
600
r
-200
1000
*-
.TE
.sp 1
.in -2
Using the default invocation, the
.B "xdelta"
command transforms the above input into the following output:
.in +2
.sp 1
.TS
l.
**Xcents
[100]
200
\.
900
-600
r
[-200]
1200
*-
.TE
.sp 1
.in -2
Notice that the input interpretation (\f(CR**cents\fR)
has been modified to \f(CR**Xcents\fR in the first record.
As can be seen, the leading or \(odoffset\(cd value `\f(CR100\fR'
has been echoed in the second record -- although it has been printed
in square brackets.
This is not a \(oddifference\(cd value since there is no previous numerical
value from which to calculate a difference;
.B "xdelta"
simply echoes the initial starting value.
The third output record contains the value `\f(CR200\fR' -- which is the
difference between the second and third input records (300 minus 100).
(Musically, we would say that the difference between 100 cents above
middle C followed by 300 cents above middle C is an increase
of plus 200 cents.)
The null-token in the fourth record has been echoed.
Null-tokens have no effect on subsequent numerical calculations and are
treated as though they are non-existent.
Thus the fifth output record contains the difference between the
third and fifth input records (1200 minus 300 equals 900).
The sixth input record (`\f(CR600\fR') is lower in value than the preceding
value (`\f(CR1200\fR') and so produces a negative output
(600 minus 1200 equals \(mi600).
The seventh input record contains no numerical value;
as a result,
.B "xdelta"
\(odbreaks\(cd operation;
it cannot calculate a numerical difference value.
The output action is to echo the input token (`\f(CRr\fR') and to begin
looking for a new offset value.
The eighth input record (`\f(CR\(mi200\fR') begins a new sequence of
numerical values; the output echoes \f(CR[\(mi200]\fR as the new offset.
The ninth input record (\f(CR1000\fR) is 1200 cents above \(mi200,
and so the corresponding output value is 1200.
.sp 1
.sp 1
Sometimes numerical values appear in tokens that the user doesn't want
processed.
A good example occurs with numbered barlines.
Consider the following simple example.
.in +2
.sp 1
.TS
l.
**dur
1.6
\(eq1
2.5
*-
.TE
.sp 1
.in -2
If the
.B "xdelta"
command is invoked with the default options, the output will be:
.in +2
.sp 1
.TS
l.
**Xdur
[1.6]
-0.6
1.5
*-
.TE
.sp 1
.in -2
In other words, the measure number (1) interacts (incorrectly)
with the duration values.
This can be avoided by using the
.B "-s"
(skip) option.
The skip option allows the user to identify records that should
not be involved in 
.I "\(*Dx"
processing.
The \f(CR**dur\fR barline signifier is the equals-sign;
hence, the command
xdelta -s = input
will cause the barlines to be ignored in the numerical calculation,
and so produce the following (correct) output:
.in +2
.sp 1
.TS
l.
**Xdur
[1.6]
\(eq1
0.9
*-
.TE
.sp 1
.in -2
Some inputs may contain multiple-stops -- that is, Humdrum data tokens
containing two or more encoded components.
The
.B "xdelta"
command is also able to process numerical data tokens containing
multiple-components.
Consider, for example, the following
**semits
file:
.in +2
.sp 1
.TS
l.
**semits
3
4 7
-3 -7 11
12
*-
.TE
.sp 1
.in -2
Notice the presence of the double- and triple-stops in the fourth and
fifth records.
Using the default invocation, the
.B "xdelta"
command processes this input as follows:
.in +2
.sp 1
.TS
l.
**Xsemits
[3]
1 4 
-7 (-11) (-14) 4 
15 19 1 
*-
.TE
.sp 1
.in -2
Once again, the input interpretation (\f(CR**semits\fR)
has been modified to \f(CR**Xsemits\fR.
The leading or offset value [3] has been echoed
in the second record.
(The user might wish to eliminate such offset values via the
.B "humsed"
command; see below.)
The third records in both the input and output contain double-stops.
In the output, the first value of the double-stop (`1') represents
the difference between 3 and 4.
The second value in the double-stop (`4') represents the difference
between 3 and 7.
In short,
.B "xdelta"
traces both possible difference \(odpaths.\(cd
In moving from the pitch D# to two concurrent pitches (E and G),
we may trace both the D#-E interval (1 semit) and the D#-G interval (3 semits).
.sp 1
.sp 1
In processing successive multiple-stops
.B "xdelta"
does not calculate all of the possible permutations.
For example, in the case of two consecutive triple-stops,
.B "xdelta"
will calculate three intervals corresponding to the first notes
in both triple-stops, the second notes, and the third notes.
.sp 1
.sp 1
Where the number of multiple-stops changes,
.B "xdelta"
operates under some special conventions.
Consider, for example, the case of a double-stop followed
by a triple-stop:
the pitches P+Q followed by X+Y+Z.
All of the possible (interval) differences might be enumerated as follows:
X-P, Y-P, Z-P, X-Q, Y-Q and Z-Q.
.B "Xdelta"
first calculates the \(odouter\(cd interval distances (X\(miP and Z\(miQ).
It then calculates a permuted set of \(odinner\(cd intervals
(Y\(miP and Y\(miQ).
The remaining intervals are considered unlikely or implausible and
are not calculated by
.B "xdelta."
.sp 1
.sp 1
In the above example,
moving from the double-stop to the triple stop between records three
and four generates the two \(odouter\(cd interval distances
(-3 minus 4 equals -7; 11 minus 7 equals 4), as well as the permuted
\(odinner\(cd intervals (-7 minus 4 equals -11; -7 minus 7 equals -14).
Both the resulting inner intervals are printed in parentheses.
A similar process occurs when moving from records four to five.
Three intervals may be traced from the 3 initial tokens to the
subsequent single token.
.sp 1
.sp 1
Depending on the goal, the presence of the parentheses makes it easy
for the user to eliminate the inner intervals using the Humdrum
stream-editor
.B "humsed."
For example, the command:
.sp 1
.sp 1
.in +2
humsed 's/([^)]*) //g' input > output
.in -2
.sp 1
.sp 1
can be used to eliminate inner intervals.
Alternatively, the command:
.sp 1
.sp 1
.in +2
humsed 's/[()]//g' input > output
.in -2
.sp 1
.sp 1
can be used to eliminate the parentheses surrounding the inner intervals.
Offset values can be transformed to null-tokens using the command:
.sp 1
.sp 1
.in +2
humsed 's/\\[[^]]*\\]/./g' input > output
.in -2
.sp 1
.sp 1
Entire records containing offset values can be eliminated using the command:
.sp 1
.sp 1
.in +2
humsed '/\\[.*\\]/d' input > output
.in -2
.sp 1
.sp 1
A further example shows how the output of
.B "xdelta"
can be recirculated as input, and the second derivative calculated.
In the example below, the first spine is the original input,
consisting of a rising-falling major arpeggio,
followed by an ascending major scale.
The second spine is the corresponding output from the command:
.sp 1
.sp 1
.in +2
xdelta -s = spine1 | humsed 's/\\[.*\\]/./' > spine2
.in -2
.sp 1
.sp 1
The original input and both outputs have been assembled together below.
.sp 1
.sp 1
Notice that barlines have been skipped and that the initial offset
value has been changed to a null token (using
.B "humsed)."
The second spine has then been used as input to
.B "xdelta"
with the result of the following command given in the third spine.
.sp 1
.sp 1
.in +2
xdelta -a -s = spine2 | humsed 's/\\[.*\\]/./' > spine3
.in -2
.sp 1
.sp 1
Notice that only absolute numerical differences have been generated in the
third spine.
In the output below, semitone pitch values are coordinated with the
interval by which it was approached (2nd spine) and by the change of
interval size (3rd spine).
Notice that large values in third spine (e.g. 10 and 6) correspond
to points in the input where the arpeggio changes direction, and
where the arpeggio changes to a scale.
(It is common to encounter such large discontinuities
whenever a pattern changes.)
.in +2
.sp 1
.TS
l l l.
**semits	**Xsemits	**XXsemits
*M3/4	*M3/4	*M3/4
\(eq1	\(eq1	\(eq1
0	.	.
4	4	.
7	3	1
\(eq2	\(eq2	\(eq2
12	5	2
7	-5	10
4	-3	2
\(eq3	\(eq3	\(eq3
0	-4	1
2	2	6
4	2	0
5	1	1
\(eq4	\(eq4	\(eq4
7	2	1
9	2	0
11	2	0
\(eq5	\(eq5	\(eq5
12	1	1
====	====	====
*-	*-	*-
.TE
.sp 1
.in -2
.in -2
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fBassemble\fR (1),
\fBhumsed\fR (1), \fBmint\fR (1),
\fBrecode\fR (1), \fBregexp\fR (1),
\fBydelta\fR (1)
.in -2


toolkit/awk/xdelta.awk
#####################################################################
##					XDELTA.AWK
# Programmed by: Tim Racinsky        Date: July, 1993
# Copyright (c) 1993 David Huron
#
# Modifications:
#   Date:      Programmer:       Description:
# June 7/94	Tim Racinsky	Modified parse_command to work with getopts
#
#
#	This program receives a valid humdrum file as input and calculates
# the changes in the numbers (the "delta x", so to speak) between records
# for each separate spine.  It is called from xdelta.ksh and is passed
# the corresponding options.
#
#	The following gives a brief summary of the available options:
#
#	-i include non-numeric data in output of processed tokens
#	-a print absolute values for output
#	-b regexp specify which tokens to stop processing on
#	-s regexp specify which tokens to ignore
#
# Main Functions used:
#	Parse_command()		Initialize_arrays()		Reset_arrays()
#	Purge_array()			Process_deltas()		Deltax()
#	Store_indicators()		Process_indicators() 	Ins_array_pos()
#	Del_array_pos()		Num_of_Exchanges()		Exchange_spines()
#	Join_paths()			Store_new_interps()		Find_number()
#
# Helper Functions used:
#	Print_indics()			Print_previous()
#	Print_current()		Print_token()
#
#	VARIABLES:
#
#  -spine_path_record:  flag to indicate if current interpretation record
#   is also a spine-path record.
#  -options:  holds the options the user selected.
#  -no_interpretation_yet:  flag to indicate if first interpretation record
#   has been parsed yet.
#  -current_no_of_spines:  holds current number of active spines.
#  -current_line:  string holding the current line of data tokens to be output.
#  -previous_max:  holds the number of 'stops' in the previous token for
#   the curent spine.
#  -number_in_previous:  flag to determine if any tokens in the previous record
#   for the current spine contained a number.
#  -current_max:  same as previous except for current token.
#  -number_in_current:  same as previous except for current token.
#  -breg_exp:  holds the regular expression that forces the program to break
#   on data tokens matching it.
#  -sreg_exp:  same as above except program 'skips' the data token.
#
#	ARRAYS:
#
#  -previous_tokens:  each element corresponds to the data token(s) for 
#   each spine of the previous record.
#  -previous_array:  when looping through each spine, this array holds the
#   tokens for the current spine that existed in the previous record (i.e. if 
#   there is a double stop then there are 2 array elements).
#  -previous_temp:  holds all non-numeric data from the array previous_array.
#   This allows only numbers to be printed or both numbers and non-number data.
#  -current_array:  same as previous_array except for current token.
#  -current_temp:  sames as preveious_temp except for current token.
#  -path_indicator:  holds spine-path indicators for current spine-path record.
#
BEGIN {
	USAGE="\nUSAGE: xdelta -h                    (Help Screen)\n       xde"\
		 "lta [-ae] [-b reg_exp] [-s reg_exp] [file ...]"
	FS = OFS = "\t"
	TRUE = 1; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	NUMBER = "[+-]?([0-9]+([.][0-9]+)?|[.][0-9]+)"
	options = breg_exp = sreg_exp ""
	previous_array[0] ;
	current_array[0] ;
	#
	# Determine the options specified by the user
	#
	parse_command()
	}
{
#
# If this is the first record of the current file, reset variables
#
if (FNR == 1)
	{
	no_interpretation_yet = TRUE
	current_no_of_spines = 0
	spine_path_record = FALSE
	initialize_arrays()
	reset_arrays()
	}
#
# Print each comment line without upsetting data values
#
if ($0 ~ /^!/) print $0
#
# If this is an interpretation record, print and process it
#
else if ($0 ~ /^\*/)
	{
	#
	# If this is the first inter. record, initialize variables
	#
	if (no_interpretation_yet)
		{
		current_no_of_spines = NF
		no_interpretation_yet = FALSE
		store_new_interps()
		initialize_arrays()
		#
		# All exclusive interpretations have an "X" prepended to them
		#
		for (i = 1; i <= current_no_of_spines; i++)
			{ if ($i ~ /^\*\*/) sub(/^\*\*/,"**X",$i) }	
		print $0
		}
	#
	# Otherwise, check to see if it is a spine-path record
	#
	else
		{
		spine_path_record = FALSE
		for (i=1; i<=current_no_of_spines ; i++)
			{			
			if ($i ~ /^\*\+$|^\*\-$|^\*\^$|^\*v$|^\*x$/) {
				spine_path_record = TRUE 
				break
				}
			}
		#
		# If it is a spine-path record, store and process the indicators
		#
		if (spine_path_record)
			{
			print $0
			store_indicators()
			process_indicators()
			if (current_no_of_spines == 0) no_interpretation_yet = TRUE
			}
		#
		# Otherwise, process the interpretations
		#
		else
			{
			store_new_interps()
			#
			# All exclusive interpretations have an "X" prepended to them
			#
			for (i = 1; i <= current_no_of_spines; i++)
				{
				if ($i ~ /^\*\*/) sub(/^\*\*/,"**X",$i)
				}	
			print $0
			}		
		}
	}
#
# Otherwise, it must be a data record
#
else
	{
	#
	# Use 'current_line' for printing the current record
	#
	current_line = ""
	#
	# Process each spine separately
	#
	for (outside = 1; outside <= current_no_of_spines; outside++)
		{
		if (options ~ /s/ && $outside ~ sreg_exp)
			{
			current_line = current_line $outside "\t"
			continue
			}
		else if (options ~ /b/ && $outside ~ breg_exp)
			{
			current_line = current_line $outside "\t"
			previous_tokens[outside] = ""
			continue
			}
		else if ($outside == ".")
			{
			current_line = current_line $outside "\t"
			continue
			}
		#
		# Reset all of the temporary arrays that are created
		#
		reset_arrays()
		#
		# Split the previous data token and the current data token by the
		# space so that multiple 'stops' may be processed separately. Also
		# purge the resulting arrays of non-numeric values and determine
		# the largest subscript in each array
		#
		previous_max = split(previous_tokens[outside],previous_array," ")
		number_in_previous = find_number(previous_array)
		purge_array(previous_array,previous_temp)
		#
		current_max = split($outside,current_array," ")
		number_in_current = find_number(current_array)
		purge_array(current_array,current_temp)
		#
		# If the previous token contains no numbers
		#
		if (!number_in_previous)
			{
			#
			# If the current token contains no numbers, just print the token
			#
			if (!number_in_current)
				current_line = current_line $outside "\t"
			#
			# Otherwise, print each numeric token in square brackets
			#
			else
				{
				for (j = 1; j <= current_max; j++)
					{
					#
					# If the user wants non-numeric data printed
					#
					if (options ~ /e/)
						{
						sub(SUBSEP,current_array[j],current_temp[j])
						if (current_array[j] != "")
							current_line = current_line "[" \
										current_temp[j] "] "
						else current_line = current_line \
										current_temp[j] " "
						}
					#
					# If the user wants only numbers printed
					#
					else
						{
						if (current_array[j] != "")
							current_line = current_line "[" \
										current_array[j] "] "
						else
							current_line = current_line \
										current_temp[j] " "
						}
					}
				current_line = substr(current_line,1\
							,length(current_line)-1) "\t"
				}
			}
		#
		# If the previous token does contain a number
		#
		else
			{
			#
			# If the current token also contains a number, process them
			#
			if (number_in_current)
				{
				process_deltas()
				}
			#
			# Otherwise, if the current token has no number, just print it
			#
			else current_line = current_line $outside "\t"
			}
		#
		# Store the current token as the previous token
		#
		previous_tokens[outside] = $outside
		}
	#
	# Print the resulting line
	#
	print substr(current_line,1,length(current_line)-1)
	}
}

############################################################################
#				FUNCTIONS USED IN XDELTA.AWK
############################################################################

############################################################################
##				Function Parse_command
# This function inspects the array ARGV in order to determine the correct
# options that the user has specified.
#
function parse_command()
	{
	if (ARGV[2] != "null") options = ARGV[2]
	if (ARGV[3] != "null")
		{
		options = options "b"
		breg_exp = ARGV[3]
		}
	if (ARGV[4] != "null")
		{
		options = options "s"
		sreg_exp = ARGV[4]
		}
	ARGV[1] = ARGV[2] = ARGV[3] = ARGV[4] = ""
	}

############################################################################
##				Function Initialize_arrays
#	This function initializes the previous_tokens to show no numbers
#
function initialize_arrays(    j)
	{
	for (j = 1; j <= current_no_of_spines; j++) previous_tokens[j] = ""
	}

########################################################################
##				Function Reset_arrays
#	This function resets the temporary arrays that are created for
# each separate spine.
#
function reset_arrays(    j)
	{
	for (j in previous_array) delete previous_array[j]
	for (j in current_array) delete current_array[j]
	for (j in previous_temp) delete previous_temp[j]
	for (j in current_temp) delete current_temp[j]
	}

######################################################################
##				Function Find_number
#	This funtion determines if the array passed to it contains any
# numbers in its elements.
#
function find_number(array,  number_found,j)
	{
	number_found = FALSE
	for (j in array)
		{
		if (array[j] ~ NUMBER)
			{
			number_found = TRUE
			break
			}
		}
	#
	# If no element was found to contain a number, return negative
	# Otherwise, return positive.
	#
	return number_found
	}

######################################################################
##				Function Purge_array
#	This function takes the array 'array' and removes all non-numeric
# material to array_temp.
#
function purge_array(array,array_temp,  j,temp)
	{
	#
	# Loop through the given array
	#
	for (j in array)
		{
		#
		# If there is no number, entire contents are transferred
		#
		if (match(array[j],NUMBER) == 0)
			{
			array_temp[j] = array[j]
			array[j] = ""
			}
		#
		# Otherwise, separate the contents
		#
		else
			{
			temp = array[j]
			array[j] = substr(array[j],RSTART,RLENGTH)
			sub(NUMBER,SUBSEP,temp)
			array_temp[j] = temp
			}
		}
	}

###################################################################
##				Function Process_deltas
#	This function determines the changes in the numbers between two
# data tokens.  Note that this may include multiple 'stops' and they
# are represented here as distinct array elements.  This function also
# continues to create 'current_line', the line that is printed to output.
# It uses the deltax() function to determine the actual token to print out
# and sends as parameters the array subscripts to process.
#
function process_deltas(    outside,inside,next_token)
	{
	#
	# If there are an equal number of 'stops' in the tokens, then match
	# them 1-1.
	#
	if (current_max == previous_max)
		{
		for (outside = 1; outside <= current_max; outside++)
			{
			next_token = deltax(outside,outside)
			current_line = current_line next_token " "
			}
		}
	#
	# If at least one token has only one number, then process that number
	# with the rest from the other token
	#
	else if (current_max == 1 || previous_max == 1)
		{
		for (outside = 1; outside <= current_max; outside++)
			{
			for (inside = 1; inside <= previous_max; inside++)
				{
				next_token = deltax(outside,inside)
				current_line = current_line next_token " "
				}
			}
		}
	#
	# Otherwise, there are 3 cases to deal with.  All of them start by 
	# processing the first two numbers in each token and finishing with
	# the last two numbers in each token. Middle values are output 
	# within parantheses.
	#
	else
		{
		current_line = current_line deltax(1,1) " "
		#
		# If the previous token had only 2 elements (Note that the current
		# token must have greater than 2 elements).
		#
		if (previous_max == 2)
			{
			for (outside = 2; outside < current_max; outside++)
				{
				for (inside = 1; inside <= previous_max; inside++)
					{
					next_token = deltax(outside,inside)
					current_line = current_line "(" next_token ") "
					}
				}
			}
		#
		# If the current token has only 2 elements (Likewise, the previous
		# token has greater than 2 elements).
		#
		else if (current_max == 2)
			{
			for (outside = 1; outside <= current_max; outside++)
				{
				for (inside = 2; inside < previous_max; inside++)
					{
					next_token = deltax(outside,inside)
					current_line = current_line "(" next_token ") "
					}
				}
			}
		#
		# Otherwise, both tokens have greater than 2 elements
		#
		else
			{
			for (outside = 2; outside < current_max; outside++)
				{
				for (inside = 2; inside < previous_max; inside++)
					{
					next_token = deltax(outside,inside)
					current_line = current_line "(" next_token ") "
					}
				}
			}
		current_line = current_line deltax(current_max,previous_max) " "
		}
	current_line = substr(current_line,1,length(current_line)-1) "\t"
	}

##########################################################################
##			 	Function Deltax
#	This function is called from Process_deltas() and is used to do the
# actual subtraction of the values.
#
function deltax(current,previous,   result,temp)
	{
	#
	# If current_array at the current position has no numeric data,
	# then simply return the non-numeric data
	#
	if (current_array[current] == "") return current_temp[current]
	#
	# Else, if previous_array at current position has no numeric data
	# 
	else if (previous_array[previous] == "")
		{
		#
		# If non-numeric data is to be output, return the original token
		# in square brackets.
		#
		if (options ~ /e/)
			{
			sub(SUBSEP,current_array[current],current_temp[current])
			return "[" current_temp[current] "]"
			}
		#
		# Otherwise, just print the numbers in square brackets.
		#
		else return "[" current_array[current] "]"
		}
	#
	# If both contain numeric data
	#
	else
		{
		result = current_array[current] - previous_array[previous] + 0
		if (options ~ /a/ && result < 0) result *= -1
		if (options ~ /e/)
			{
			temp = current_temp[current]
			sub(SUBSEP,result,temp)
			return temp
			}
		else return (result)
		}
	}

##############################################################
##			Function Store_indicators
#	This function allows the spine-path indicators for the 
# current record to be stored in the array 'path_indicator' so
# that they may be used later.
#
function store_indicators(   i)
	{
	for (i = 1; i <= current_no_of_spines; i++)
		{
		if ($i == "*") path_indicator[i] = "*"
		else path_indicator[i] = substr($i,2,1)
		}
	}

##################################################################
##			Function Process_indicators
#	This function takes the spine-path indicators that were stored
# in the array 'path_indicator' in the function 'store_indicators'
# and manipulates the arrays 'path_indicator' and 
# according to the contents of the array 'path_indicator'.
#
function process_indicators(    i)
	{
	i = 1
	while (i <= current_no_of_spines)
		{
		#
		# For split-paths, both of the resulting paths use the value of
		# the path before the split
		#
		if (path_indicator[i] == "^")
			{
			ins_array_pos(i+1)
			path_indicator[i+1] = "*"
			previous_tokens[i+1] = previous_tokens[i]
			}
		#
		# For spine termination, terminate everything for that path
		#
		if (path_indicator[i] == "-")
			{
			del_array_pos(i)
			i--
			}
		#
		# For exchange paths, exchange the previous values of the paths
		#
		if (path_indicator[i] == "x")
			{
			spines_to_exchange = ""
			num_of_exchanges()
			exchange_spines()
			}
		#
		# For join paths, join the values of the original paths to
		# form a multiple 'stop' token.
		#
		if (path_indicator[i] == "v")
			{
			join_paths(i,i+1)
			del_array_pos(i+1)
			if (path_indicator[i+1] != "v")
				{
				path_indicator[i] = "*"
				join = FALSE
				}
			else i--
			}
		#
		# For a new path, start the path as fresh
		#
		if (path_indicator[i] == "+")
			{
			ins_array_pos(i+1)
			path_indicator[i+1] = "*"
			previous_tokens[i+1] = ""
			}
		i++
		} #end while loop
	} #end function process_indicators

############################################################
##			Function Ins_array_pos
#	This function inserts new positions in the arrays 'path_indicator'
# and and 'previous_tokens' copies elements so that everything is preserved
#
function ins_array_pos(current_element,     j) 
	{
	for (j = current_no_of_spines+1; j > current_element; j--)
		{
		path_indicator[j] = path_indicator[j-1]
		previous_tokens[j] = previous_tokens[j-1]
		}
	path_indicator[current_element-1] = "*"
	current_no_of_spines++
	}

##############################################################
##			Function Del_array_pos
#	Performs the opposite of function 'ins_array_pos'.
#
function del_array_pos(current_element,     j)
	{
	for (j = current_element; j < current_no_of_spines; j++)
		{
		path_indicator[j] = path_indicator[j+1]
		previous_tokens[j] = previous_tokens[j+1]
		}
	delete path_indicator[j]
	delete previous_tokens[j]
	current_no_of_spines--
	} 

###############################################################
##			Function Num_of_exchanges
#	This function determines how  many exchange-path indicators
# appear in the current spine-path record and stores them in
# the variable 'spines_to_exchange'.
#
function num_of_exchanges(    j)
	{
	for (j = 1; j <= current_no_of_spines; j++)
		{
		if (path_indicator[j] == "x")
			{
			spines_to_exchange = spines_to_exchange j "\t"
			path_indicator[j] = "*"
			}
		}
	return
	}	

###################################################################
##				Function Exchange_spines
#	This function exhanges two spines by exchanging the corresponding
# elements in 'previous_tokens'.
#
function exchange_spines(    arraya,temp)
	{
	split(spines_to_exchange,arraya)
	temp = previous_tokens[arraya[1]]
	previous_tokens[arraya[1]] = previous_tokens[arraya[2]]
	previous_tokens[arraya[2]] = temp
	}

############################################################
##				Function Join_paths
#	This function appends tokens of the joined spines to form a single
# token as a multiple stopped token.
#
function join_paths(first_spine,second_spine,     i,j)
	{
	previous_tokens[first_spine] = previous_tokens[first_spine] \
				" " previous_tokens[second_spine]
	}

################################################################
##				Function Store_new_interps
#	This function resets a spine if an exclusive interpretation is found.
#
function store_new_interps(      j)
	{
	for (j = 1; j <= current_no_of_spines; j++)
		{
		if ($j ~ /^\*\*/) previous_tokens[j] = ""
		}
	}

#####################################################################
##			Helper Function Print_indics
#
function print_indics(        j)
	{
	print "Array Path_Indicator"
	for (j in path_indicator) print j, path_indicator[j]
	print "End Array Path_Indicator"
	}

######################################################################
##			Helper Function Print_previous
#
function print_previous(   j)
	{
	print "Array Previous_array"
	for (j in previous_array) print j, previous_array[j]
	print "End Array Previous_array"
	}

######################################################################
##			Helper Function Print_current
#
function print_current(   j)
	{
	print "Array Current_array"
	for (j in current_array) print j, current_array[j]
	print "End Array Current_array"
	}

######################################################################
##			Helper Function Print_token
#
function print_token(element)
	{
	print "Previous_tokens element"
	print element, previous_tokens[element]
	print "End Previous_tokens element"
	}
